File: listnode/234_palindrome_linked_list.py
from ListNode import ListNode


class Solution:

    def isPalindrome(self, head: ListNode):
        """
        Reverse linked list first, then compare with original
        """
        if not head:
            return True 
        ori = []
        def help(h):
            ori.append(h.val)
            if not h.next:
                return h
            last = help(h.next)
            h.next.next = h
            h.next = None
            return last
        
        re = help(head)
        while re:
            if re.val != ori[0]:
                return False
            re = re.next
            ori = ori[1:]
        return True

    def isPalindrome_2(self, head: ListNode):
        left = head

        def help(right: ListNode):
            """
            Postorder traversal, using two pointers
            """
            nonlocal left
            if right is None:
                return True
            res = help(right.next)
            print(f' {left.val, right.val}')
            res = res and (left.val == right.val)
            left = left.next
            return res

        return help(head)

    def isPalindrome_3(self, head: ListNode):
        slow = fast = head
        # get mid
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        # odd situation, need to move one more step
        if not fast is None:
            slow = slow.next
        
        def help(node):
            pre = None
            cur = node
            while not cur is None:
                cur.next, cur, pre = pre, cur.next, cur
            return pre
        
        left = help(slow)
        right = head
        while right and left:
            if right.val != left.val:
                return False
            right = right.next
            left = left.next
        return True


nums = [1, 1, 2, 1]
nums = [1, 1, 1]
nodes = [ListNode(val=a) for a in nums]
for i in range(len(nodes)-1):
    nodes[i].next = nodes[i+1]
head = nodes[0]

test = Solution()
# res = test.isPalindrome(head)
# res = test.isPalindrome_2(head)
res = test.isPalindrome_3(head)
print(res)

File: c++/234_palindrome_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* temp;
    bool isPalindrome(ListNode* head) {
        temp = head;
        return check(head);
    }
    bool check(ListNode* p) {
        if (NULL == p) return true;
        bool res = check(p->next) & (p->val == temp->val);
        temp = temp->next;
        return res;
    }

    ListNode* reverse(ListNode* p) {
        // reverse a linked list
        if (!p) return NULL;
        ListNode* pre = NULL;
        ListNode* temp = p;
        while (p) {
            temp = p->next;
            p->next = pre;
            pre = p;
            p = temp;
        }
        return pre;
    }

    bool isPalindrome2(ListNode* root) {
        if (!root) return true;
        ListNode* slow = root;
        ListNode* fast = root;
        while (fast->next && fast->next->next) {
            fast = fast->next->next;
            slow = slow->next;
        }
        // reverse the second half
        slow->next = reverse(slow->next);
        slow = slow->next;
        while (slow) {
            if (slow->val != root->val) {
                return false;
            }
            else {
                slow = slow->next;
                root = root->next;
            }
        return true;

    }

    bool isPalindrome3(ListNode* root) {
        if (!root) return true;
        ListNode* fast = root;
        ListNode* slow = root;
        ListNode* rev = NULL;
        ListNode* temp = NULL;
        ListNode* temp2 = NULL;

        while (fast && fast->next) {
            fast = fast->next->next;
            // reverse the first half of linked list, be careful!
            temp = slow->next;
            temp2 = rev;
            rev = slow;
            rev.next = temp2;
            slow = temp;
        }
        if (fast) {
            slow = slow->next;
        }
        while (rev) {
            if (rev->val != slow->val) return false;
            rev = rev->next;
            slow = slow->next;
        }
        return true;
    }
};